packages/gen-ai-hub/src/client/openai/openai-client.ts

```diff
@@ -1,36 +1,43 @@
+import { HttpRequestConfig } from '@sap-cloud-sdk/http-client';
 import {
-  BaseLlmParameters,
   CustomRequestConfig,
-  executeRequest
+  DeploymentResolver,
+  executeRequest,
+  resolveDeployment
 } from '../../core/index.js';
-import { BaseClient } from '../interface.js';
 import {
   OpenAiChatCompletionParameters,
   OpenAiEmbeddingParameters,
   OpenAiEmbeddingOutput,
-  OpenAiChatCompletionOutput
+  OpenAiChatCompletionOutput,
+  OpenAiChatModel,
+  OpenAiEmbeddingModel
 } from './openai-types.js';
+import { get } from 'http';
 
 const apiVersion = '2024-02-01';
 
 /**
  * OpenAI GPT Client.
  */
-export class OpenAiClient implements BaseClient<BaseLlmParameters> {
+export class OpenAiClient {
   /**
    * Creates a completion for the chat messages.
    * @param data - The input parameters for the chat completion.
    * @param requestConfig - The request configuration.
    * @returns The completion result.
    */
   async chatCompletion(
+    model: OpenAiChatModel,
     data: OpenAiChatCompletionParameters,
+    deploymentResolver: DeploymentResolver = getDeploymentResolver(model),
```

Comment:
[req] Previously we discussed the idea of a deployment configuration, where the user could set different parameters to identify the deployment, e.g. `{ deploymentId: 'xx'}` or `{ modelName: 'xx' }`. I think this was quite a good idea as it is simply and clear, but not too verbose.
What I like about your proposal is, that the model for the `OpenAiClient` is restricted to OpenAI models only, which makes sense and I would like to keep that.
However, I don't like:
1. That users have to specify a constant to identify the model. I would prefer a string literal union type, so that they don't need to import anything additional. I guess the versioning might affect the design here.
2. The separation between model and deployment resolver. This is not 100% clear to me yet, but semantically I don't think the separation is needed, because the user provides this for the same purpose. I think the original idea of the deployment configuration covers this in a cleaner way. It could even be extended by a deployment resolver, however, I wonder whether this isn't premature. Are we aware of any use cases, where users would want to overwrite the resolution behavior?

Reply:
Good points, I also asked myself these two questions when implementing this.

(1) The reason I went with a complex type to represent the models is that it would be useful to have more information about them, e.g. if they are text/embedding/image models, which versions they have or if they are multi-modal. Maybe more information in the future, e.g. if they allow for other specific features like [response format](https://platform.openai.com/docs/api-reference/chat/create#chat-create-response_format). 

However, I'm not sure yet which aspects we want to be constant, and where we would need to have dynamic information. E.g., if we want to allow the user to select a specific version of a model, we either need (modelName * version) constants, or some constructor / function to create an object dynamically. 

Finally, I'd agree that string is simpler. I don't know how common enum or enum-like style is in TS. If we prefer, we can stick to string for simplicity or alternatively allow both options somehow.

(2) I split model and deployment as I see them as conceptually different things: A model is a description of a foundation model and its capabilities (e.g. versions, parameter options etc.). A deployment is a connectivity detail of AI Core (has a URL, is linked to a configuration, scenario, resource group).

So far it just so happens that only one of the two is needed for the basic OpenAI related functionality, since we currently only care about the deployment id. But already for Orchestration the model is part of the payload and not tied to any deployment. Also, a deployment can be changed at runtime to host different models.

Separating the two allows us to evolve the connectivity aspects independently from another. For example, we could (and probably should in some way) re-use the model constants in the Orchestration service API, assisting users with which models are available.

Still, it should be as convenient as possible for the user. What I liked about the current approach is that it nudges users into using the convenience functionality, and away from hard-coding GUIDs into their application code. Maybe this is also possible with other approaches. I am curious what else we will come up with 😉